cpp/plugins/cucim.kit.cuslide2/src/cuslide/nvimgcodec/nvimgcodec_decoder.cpp

```diff
@@ -0,0 +1,454 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "nvimgcodec_decoder.h"
+#include "nvimgcodec_tiff_parser.h"
+
+#ifdef CUCIM_HAS_NVIMGCODEC
+#include <nvimgcodec.h>
+#endif
+
+#include <cucim/memory/memory_manager.h>
+#include <memory>
+#include <vector>
+#include <cstring>
+#include <string>
+#include <stdexcept>
+#include <vector>
+#include <unistd.h>
+#include <mutex>
+#include <fmt/format.h>
+
+#ifdef CUCIM_HAS_NVIMGCODEC
+#include <cuda_runtime.h>
+#endif
+
+namespace cuslide2::nvimgcodec
+{
+
+#ifdef CUCIM_HAS_NVIMGCODEC
+
+// ============================================================================
+// RAII Helpers for nvImageCodec Resources
+// ============================================================================
+
+// RAII wrapper for nvimgcodecCodeStream_t
+struct CodeStreamDeleter
+{
+    void operator()(nvimgcodecCodeStream_t stream) const
+    {
+        if (stream) nvimgcodecCodeStreamDestroy(stream);
```

Comment:
Recognize this is valid, but should we use `{}`s just to be safe in case other code is added here?

```suggestion
        if (stream) { nvimgcodecCodeStreamDestroy(stream); }
```